[WB-73] Added dao for handling querying and storage of npm deps

ctdio opened the pull request.

ctdio committed:
[WB-73] Added dao for handling querying and storage of npm deps

Druotic commented on src/dao/dependency/npm.js:
```diff
@@ -0,0 +1,80 @@
+ * the query to be malformed
+ */
+function _jsonKeyExistsIn (knex, key, field) {
+  return knex.raw(`${field}::jsonb ->> ? is not null`, [ key ])
```
Is this type cast really necessary? Would simply ```knex.raw(`${field} ->> ? is not null`, [ key ])``` work as well?

On a side note, it looks like knex's query building support around json is a bit lacking :(

charlieduong94 replied:
Nope not needed, sorry thought that I needed to be explicit

austinkelleher commented on src/dao/dependency/npm.js:
```diff
@@ -0,0 +1,80 @@
+
+function _formatDependencies (data) {
+  return {
+    dependencies: JSON.stringify(data.dependencies || {}),
```
Should we just store a null (not sure what the right term is in postgres land) if these are empty? We are going to have a lot of fields that simply store `{}` and never a actually store anything else.

charlieduong94 replied:
Storing `null` or `undefined` is fine. Will change.

charlieduong94 commented on src/models/dependencies/Npm.js:
```diff
@@ -0,0 +1,27 @@
+/**
+* Model for Installations stored in the databases
```
whoops will fix

charlieduong94 commented on src/dao/migrations/20171114125010_create-npm-dependencies-table.js:
```diff
@@ -0,0 +1,27 @@
+
+exports.up = function ({ schema }) {
+  logger.info('Attempting to run "up" on "create-repository-table"')
+  logger.info('Attempting to create repository table with table name: ', REPOSITORY_TABLE_NAME)
```
will fix

charlieduong94 commented on src/dao/dependency/npm.js:
```diff
@@ -0,0 +1,80 @@
+    .where(keyExistsIn('dependencies'))
+    .orWhere(keyExistsIn('dev_dependencies'))
+    .orWhere(keyExistsIn('peer_dependencies'))
+    .orWhere(keyExistsIn('optional_dependencies'))
```
Yeah, do have some worries about this as well, but I feel like we should optimize this later once we actually reach that sort of scale. This _should_ still be much faster than performing something like `couchdb`'s design functions.

ctdio committed:
[WB-73] Minor clean up

Removed jsonb typecasting in dependency query

Allow for null or undefined data to be placed in npm deps json fields

Fixed copypasta'd comments
